pkg/datasource/accessors.go

```diff
@@ -0,0 +1,182 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datasource
+
+import (
+	"errors"
+	"fmt"
+	"slices"
+)
+
+// FieldAccessor grants access to the underlying buffer of a field
+type FieldAccessor interface {
+	Name() string
+
+	// Size returns the expected size of the underlying field or zero, if the field has a dynamic size
+	Size() uint32
+
+	// Get returns the underlying memory of the field
+	Get(data Data) []byte
+
+	// Set sets value as the new reference for the field; if the FieldAccessor is used for the member of a
+	// statically sized payload (for example a member of an eBPF struct), value will be copied to the existin
+	// memory instead.
+	Set(data Data, value []byte) error
+
+	// IsRequested returns whether the consumer is interested in this field; if not, operators are not required
+	// to will them out
+	IsRequested() bool
+
+	// AddSubField adds a new field as member of the current field; be careful when doing this on an existing
+	// non-empty field, as that might be dropped on serialization // TODO
+	AddSubField(name string, opts ...FieldOption) (FieldAccessor, error)
+
+	// GetSubFieldsWithTag returns all SubFields matching all given tags
+	GetSubFieldsWithTag(tag ...string) []FieldAccessor
+
+	// Parent returns the parent of this field, if this field is a SubField
+	Parent() FieldAccessor
+
+	// SubFields returns all existing SubFields of the current field
+	SubFields() []FieldAccessor
+
+	// SetHidden marks a field as hidden (by default) - it can still be requested
+	SetHidden(bool)
+}
+
+type fieldAccessor struct {
+	ds *dataSource
+	f  *field
+}
+
+func (a *fieldAccessor) Name() string {
+	return a.f.Name
+}
+
+func (a *fieldAccessor) Size() uint32 {
+	return a.f.Size
+}
+
+func (a *fieldAccessor) Get(d Data) []byte {
+	if FieldFlagEmpty.In(a.f.Flags) {
+		return nil
+	}
+	if a.f.Size > 0 {
+		// size and offset must be valid here; checks take place on initialization
+		return d.(*data).Payload[a.f.PayloadIndex][a.f.Offs : a.f.Offs+a.f.Size]
+	}
+	return d.(*data).Payload[a.f.PayloadIndex]
+}
+
+func (a *fieldAccessor) SetHidden(val bool) {
+	a.ds.lock.Lock()
+	defer a.ds.lock.Unlock()
+	if !val {
+		FieldFlagHidden.RemoveFrom(&a.f.Flags)
+	} else {
+		FieldFlagHidden.AddTo(&a.f.Flags)
+	}
+}
+
+func (a *fieldAccessor) Set(d Data, b []byte) error {
+	if FieldFlagEmpty.In(a.f.Flags) {
+		return errors.New("field cannot contain a value")
+	}
+	if a.f.Size != 0 && uint32(len(b)) != a.f.Size {
+		return fmt.Errorf("invalid size, expected %d, got %d", a.f.Size, len(b))
+	}
```

Comment:
This test will cause issues with gadgets that send a variable event size, like trace exec.


```bash
$ sudo -E ig run-experimental trace_exec
INFO[0000] Experimental features enabled                
WARN[0000] unknown type for variable "gadget_filter_by_mntns": Var:"gadget_filter_by_mntns" 
WARN[0000] unknown type for variable "gadget_filter_by_mntns": Var:"gadget_filter_by_mntns" 
PID    PPID   UID    GID    LOGIN… SESSI… RETV… ARGS… ARGS… COMM     ARGS     TIMESTA… K8S.NODE K8S.NAM… K8S.POD… RUNTIME… RUNTIME… RUNTIME… RUNTIME… RUNTIME…
INFO[0005] error setting buffer: invalid size, expected 7752, got 92 
INFO[0007] error setting buffer: invalid size, expected 7752, got 92 
INFO[0008] error setting buffer: invalid size, expected 7752, got 92
```

I suppose checking that the size isn't bigger that total size will be enough.

(for sure we can discuss all this "variable event size" approach layer on, but I'd like to avoid this discussion on this PR)